### ISSUE TITLE
Serde Support

### ISSUE THREAD

**Issue by scott-wilson** [kdl-org/kdl-rs#71]
This is another crack at Serde support (linking #17 for historical reference).

The basic API that I'm thinking of would look something like this (borrowing from Serde JSON):

```rust
#[derive(Serialize, Deserialize)]
struct Config {
    root_dir: PathBuf,
    iterations: u32,
}

// Using KDL v1 syntax
let data = r#"
config {
    root_dir "/path/to/project"
    iterations: 3
}
"#;

// Deserialize from string
let conf: Config = kdl::from_str(data).unwrap();

// Deserialize from file
let mut file = std::fs::File::open("config.kdl").unwrap();
let conf2: Config = kdl::read(&mut file).unwrap()

// Serialize to string
let roundtrip_data = kdl::to_string(&conf).unwrap();

// Serialize to file
let mut file = std::fs::File::create("config.kdl").unwrap();
kdl::write(&mut file, &conf).unwrap();
```

If no one has an issue, I'd be happy to take this on.

**Comment by zkat** [comment]
Two things:

1. we probably want to look at what the serde xml parsers do here, because children can mean different things and it needs to be configurable
2. We should wait until #70 lands to do this, since that’s going to break a lot of stuff

**Comment by scott-wilson** [comment]
> Two things:
> 
> 1. we probably want to look at what the serde xml parsers do here, because children can mean different things and it needs to be configurable
> 2. We should wait until #70 lands to do this, since that’s going to break a lot of stuff

1. Sounds good, I'll do some research.
2. Sounds good. I may do a POC anyways to get a feel for how to get it done. Worst case, it gets scraped and I use the knowledge to do something better.

**Comment by zkat** [comment]
That seems fine by me. I expect it to be a bit of a challenge tbh but that might be because I’ve never implemented a serde thing before.

**Comment by scott-wilson** [comment]
Same, haven't implemented serde before. But, that's the fun! We'll learn how to implement it together!

**Comment by zkat** [comment]
Another useful reference might be Kaydle: https://github.com/Lucretiel/kaydle, which is the last attempt I know of of someone trying to serdeify kdl, and they did some interesting things, just never finished it (it's a big project, it turns out!)

**Comment by zkat** [comment]
although part of the reason Kaydle was big was because it wrote its own parser and it was being fancy about it. With kdl-rs you'd only really need to convert the `KdlDocument` struct through serde.

**Comment by zkat** [comment]
sigh. I think this is going to need a completely new parser. Can I just catch a break lmao.

**Comment by zkat** [comment]
@TheLostLambda any news on this, btw?

**Comment by moverval** [comment]
Would be cool to see serde support. Right now it is only possible to add kdl as a hard dependency, serde would abstract much complexity away and would even make the code faster by giving more restrictions by how parsing works.

The read and write functions are perfect. I am looking forward seeing this features!

**Comment by zkat** [comment]
The way it's looking right now, we might end up going with [facet instead of serde](https://github.com/facet-rs/facet-kdl). It wouldn't be as tightly integrated into the serde ecosystem, but [facet is really cool](https://facet.rs/) for many reasons, including offering much faster compile times, so I think this would be a win. I'm also open to considering someone adding serde-proper support in the future, behind a non-default flag, but I'd like to see how far facet can get us, first.

**Comment by Lord-Valen** [comment]
Since facet-kdl has been discontinued, what now? Is serde or facet support on the menu?

**Comment by philocalyst** [comment]
I would contribute this :)

**Comment by AltriusRS** [comment]
> Since facet-kdl has been discontinued, what now? Is serde or facet support on the menu?

@Lord-Valen Looking into the (very little) public context ([Link 1](https://github.com/facet-rs/facet/issues/1727), [link 2](https://github.com/facet-rs/facet/issues/1934#issuecomment-3801153312)) on _why_ facet-kdl was discontinued, I doubt we will be seeing facet support.

I do wish we could have nice things but it seems that would be too difficult in the modern world. Additionally, based on it having been six months since this issue had any progress, I haven't got any hope for serde either. Hopefully someone will maintain the existing facet-kdl implementation, unfortunately that somebody won't be me, I haven't the time.

I do hope that @zkat has something to say regarding the state of this issue and where it lies.

**Comment by philocalyst** [comment]
I don't know if @zkat has anything to say, but I have an active PR for Serde! Please contribute if you'd like #152

▰▰▰▰▰
Miles Wirth 🙃

**Comment by computerquip-work** [comment]
Since I can't find context anywhere on these issues despite it being relevant as to wth is going on:

Basically, there's this almost-extremist anti-LLM movement that basically believes that if you so much as incorporate, use, or even talk nicely about AI, then you're a slop programmer. There was a repository hosted on codeberg that contained a public list of "slop", including programmers and projects that use AI.

zkat had kindly added the facet author as a "slop" programmer, opening him up for public harassment. This appears to have given a lot of people the "ick" in regard to zkat as a person.

To put into perspective how ridiculous this list was, they eventually even started adding things like `git` and codeberg, the hosting site that the list itself was hosted on. Then it was taken down because it turned into a public joke despite the original serious intentions.

EDIT: Also, I just realized there are forks. Here's the commit made by zkat: https://codeberg.org/small-hack/open-slopware/commit/8277d5c6c1f11015e108e599e56293f3adde9658

**Comment by philocalyst** [comment]
Interesting, but not really relevant to the issue (For anyone who wants more information, search around "openslopware"), but please be nice to the maintainers here, there's really no place for this in OSS, it's all volunteer work.

I do think this can *mostly* be closed, anyways @zkat

**Comment by computerquip-work** [comment]
It is relevant because there is otherwise no explanation at all as to why facet-kdl is dead despite it being noted as a possible point of focus, in addition to understanding the context of why this issue is in the state it's in and how to set expectations for the outcome of this issue.

**Comment by philocalyst** [comment]
My PR was merged...

▰▰▰▰▰
Miles Wirth 🙃

**Comment by computerquip-work** [comment]
Your PR doesn't contain facet support is my point, despite the authors last message being that facet would be a focus instead.

**Comment by computerquip-work** [comment]
I'm also not clear if serde support is going to stay since zkat has labelled it as AI slop.

**Comment by zkat** [comment]
@computerquip-work please keep conversations on-topic and don't give me an excuse to block you.

**Comment by zkat** [comment]
To be clear: I landed the serde support PR, so that's what we'll be leaning on. Facet support is no longer planned, and that's not open to further discussion at this time.

**Comment by AltriusRS** [comment]
Obviously, trying to keep this respectful, but words are hard and I often suck with them so apologies in advance.

Was it really a spat about the use of AI assistance that resulted in the Rust community having to suffer the loss of quite possibly the best option for actually handling KDL (respectfully, the ergonomics of using the first party package are.... they leave a lot to be desired is all I will say.). If so then sheesh. 

I was really hopefully for KDL. I'll be looking for something else to use in future instead now. Didn't realise that even whispering the first and seventh letters of the alphabet could get me put on a hit list.

If @zkat would like to clarify, my emails are open and listed on my profile, feel free to reach out because I am honestly looking for people to be adults and work through their issues with each other here. (hint hint Kat.) But, if you are completely against the existence of LLMs, which is what they are. there is no such thing as "artificial intelligence" just a fancy block of if statements with math involved. I do understand, but want to have a genuine conversation with you about the use of such tools and how they can be used to help people like myself (with chronic health issues and multiple disabilities which affect my ability to actually do things myself for example.)

I know my comment is off-topic and apologise to everyone except @zkat for the inevitable emails this will generate.  Wish you all the best.

PS: The IntelliJ plugin needs updating. It's not a good look that the first party extension doesn't support the current version of the language itself. I was going to submit a patch but I don't know kotlin and you seem to hate the existence of the tool that would actually enable me to contribute at all, sooo I guess I'll just not.

**Comment by zkat** [comment]
@AltriusRS I think when someone sets a boundary about discussing a topic, and even takes action on enforcing that boundary, it's not a good look to step in and say "I get that you have a boundary, but I wish to cross it anyway." Please stop.

**Comment by zkat** [comment]
For those interested in actually moving this issue forward: I am a maintainer, and I do this in my spare time, based on my ability and interest at any given point in time. I'm not obligated or paid to do this, nor have I asked anyone to pay me for this work. If you are interested in moving it forward, I won't block you from it (provided your contribution is human-authored), and I am willing to take time to review/merge/release the feature once it's ready. For the time being, I don't have a timeline on when I will be able to add the last couple of details to the existing PR myself.

**Comment by AltriusRS** [comment]
Your interpretation was different to my own, clearly. I wanted a civil discussion on AI. Not facet. But your doubling down instead of actually trying to address the issue I raised concern about has only solidified my own opinions. Good luck in future. I wish you all the best and truly hope that you do eventually get to choose to engage with the rust community properly, instead of being the way you currently are.

I do not speak for us all, nor will I ever claim to. But I will personally never choose projects which actively hinder their own success by doubling down on just banning things which need nuanced discussion. 

Good luck, and goodbye.

**Comment by zkat** [comment]
@AltriusRS This was never a forum for this discussion. It was never open to it. Go somewhere else if you want to have a "discussion". I made that very clear.
